nuxeo-features/nuxeo-platform-csv-export/src/main/java/org/nuxeo/ecm/platform/csv/export/io/DocumentPropertyCSVWriter.java

```diff
@@ -0,0 +1,153 @@
+import java.util.List;
+import java.util.Locale;
+import java.util.stream.Collectors;
+
+import org.apache.commons.csv.CSVPrinter;
+import org.nuxeo.common.utils.i18n.I18NUtils;
+import org.nuxeo.ecm.core.api.model.Property;
+import org.nuxeo.ecm.core.api.model.impl.ArrayProperty;
+import org.nuxeo.ecm.core.io.marshallers.csv.AbstractCSVWriter;
+import org.nuxeo.ecm.core.io.registry.reflect.Setup;
+import org.nuxeo.ecm.core.schema.types.ListType;
+import org.nuxeo.ecm.core.schema.types.Type;
+import org.nuxeo.ecm.core.schema.types.primitives.BinaryType;
+import org.nuxeo.ecm.core.schema.types.resolver.ObjectResolver;
+import org.nuxeo.ecm.directory.Directory;
+import org.nuxeo.ecm.directory.DirectoryEntryResolver;
+import org.nuxeo.ecm.directory.Session;
+
+/**
+ * @since 10.3
+ */
+@Setup(mode = SINGLETON, priority = REFERENCE)
+public class DocumentPropertyCSVWriter extends AbstractCSVWriter<Property> {
+
+    public static final String LIST_DELIMITER = "\n";
+
+    public static final String LANG_CTX_DATA = "lang";
+
+    public DocumentPropertyCSVWriter() {
+        super();
+    }
+
+    @Override
+    protected void write(Property entity, CSVPrinter printer) throws IOException {
+        if (entity.isScalar()) {
+            writeScalarProperty(entity, printer);
+        } else if (entity.isList()) {
+            writeListProperty(entity, printer);
+        } else {
+            writeUnsupported(entity.getType(), printer);
+        }
+    }
+
+    @Override
+    protected void writeHeader(Property entity, CSVPrinter printer) throws IOException {
+        printer.printRecord(entity.getXPath());
+    }
+
+    protected void writeScalarProperty(Property entity, CSVPrinter printer) throws IOException {
+        Object value = entity.getValue();
+        Type type = entity.getType();
+        if (type instanceof BinaryType) {
+            writeUnsupported(type, printer);
+        } else {
+            String valueAsString = null;
+            if (value == null) {
+                printer.print(null);
+            } else {
+                valueAsString = type.encode(value);
+                printer.print(valueAsString);
+            }
+            ObjectResolver resolver = entity.getType().getObjectResolver();
+            if (resolver instanceof DirectoryEntryResolver) {
+                DirectoryEntryResolver directoryEntryResolver = (DirectoryEntryResolver) resolver;
+                Directory directory = directoryEntryResolver.getDirectory();
+                if (VOCABULARY_TYPES.contains(directory.getSchema())) {
+                    if (value == null) {
+                        printer.print(null);
+                    } else {
+                        try (Session session = directory.getSession()) {
+                            String lang = ctx.getParameter(LANG_CTX_DATA);
+                            String label = (String) session.getEntry(valueAsString).getProperty(directory.getSchema(),
+                                    "label");
```

Comment:
Treat the case where `getEntry` returns `null` (non-existing or no read access to the directory)